src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(value);
+        if (size == 0) {
+            head = tail = node;
+            node.prev = null;
+            node.next = null;
+        } else {
+            tail.next = node;
+            node.prev = tail;
+            tail = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddedIndex(index);
+        Node<T> newNode = new Node<>(value);
+        if (head == null && tail == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else if (index == size) {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> currentNode = getNodeByIndex(index);
+            newNode.next = currentNode;
+            newNode.prev = currentNode.prev;
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T elements : list) {
+            add(elements);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T result = currentNode.value;
+        currentNode.value = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        if (index == 0) {
+            T removedElement = head.value;
+            head = head.next;
+            size--;
+            return removedElement;
+        } else if (index == size - 1) {
+            T removedElement = (T) tail.value;
+            tail = tail.prev;
+            size--;
+            return removedElement;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int index = getIndexByNode(object);
+        if (index == -1) {
+            return false;
+        } else {
+            remove(index);
+            return true;
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Index out of bounds");
```

Comment:
```suggestion
            throw new ArrayIndexOutOfBoundsException("Index" + index + " out of bounds");
```
Use more informative message